README
README file

Note:
fpop: f(9 directions)


Build for original case
    mkdir build_origin
    cd build_origin
    cmake .. -DADD_FLAGS='-O3'
    make

stream immediately after collision
Build with overlap
    streaming after each collision
    mkdir build_combine
    cd buil_overlap
    cmake .. -DADD_FLAGS='-O3 -DCOMBINE'
    make

bound, not important
Build with quicktest
    for two iteration in one time
    mkdir build_quicktest
    cd build_quicktest
    cmake .. -DADD_FLAGS='-O3 -DQUICKTEST'
    make

# final version
Build with tight
    increase overlap with proposed method
    mkdir build_tight
    cd build_tight
    cmake .. -DADD_FLAGS='-O3 -DTIGHT'
    make

Build with tight_block
    increase overlap with proposed method
    mkdir build_tight_block
    cd build_tight_block
    cmake .. -DADD_FLAGS='-O3 -DTIGHT_BLOCK'
    make

# Yuankun version
Build with segment
    increase overlap with Yuankun's method
    mkdir build_segment
    cd build_segment
    cmake .. -DADD_FLAGS='-O3 -DSEGMENT'
    make


changelog
# Change Log
All notable changes to this project will be documented in this file.
The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

##[0.3.0] 2018 Feb 13
origin, combine, tight, tight_block, panel
All have its own correct OpenMP implementation

##[0.2.0] 2018 Jan 30
#### Prepare to add OpenMP
https://github.iu.edu/IUPUI-CS-HPC/ScalableIB/blob/master/ParallelCode/src_v2_noncubeOpenMP/lbm_ib_omp.c

##[0.2.0] 2017 Nov 30
#### added
Correct insert updateZeroGradientBoundary into collision function for all version
-- After 1st collision&streaming on [iX][iY], Note that [iX-1][iY-1] is ready for second compuation but first store their rho.
#### Simple Optimization
Remove if in computing the second collision&streaming for upper boundary

##[0.1.9] 2017 Nov 27
#### added
Try to insert updateZeroGradientBoundary

##[0.1.9] 2017 Nov 20
#### added
Line lx, ly compuatation and results are verified correctly
Remove periodic boundary condition.

##[0.1.8] 2017 Nov 19
lx*ly
4096*4096, blk_size=32, segment=512
        origin    overlap   tight(Block)    tight(segment)
        116.4s      76.6s      70.1s          69.8

81920*32, step=100
Tune blk_size
        origin overlap tight
        17.9   13.3   10.5

32*81920, step=100
Tune blk_size
        origin overlap tight
        21   15.5   12.7


##[0.1.8] 2017 Nov 17
4096*4096, blk_size=128
        origin overlap quicktest tight(Block) tight(segment)
        112.5s   71s     62.04s        69s     125.59s
#### Reason
compute from Y direction, block size=128 is too big





##[0.1.8] 2017 April 26
#### added
##[0.1.7] 2017 April 24
####added
-detailed results are saved in excel files.
####todo
-assembly code to explain performance difference

##[0.1.6] 2017 April 20
change the radio of obstacole from y/10 to x/10

#### adjusted based on column major
        tight has better results because the offset?
lx =50

ly=20000, maxT=100
        origin overlap quicktest tight
        5.90s   4.45s     4.45s     4.25s
ly=200, maxT=1000

        origin overlap quicktest tight
        0.58s   0.520s     0.514s  0.50



## [0.1.5] 2017 April 18
#### tight overlapped, 6.7s , slower!
    lx =250, ly=50, maxT=10000
        origin overlap quicktest tight
        6.3s   5.6s     5.6s     6.7s

#### configuration
    lx=20000, ly=50, maxT=100, 20000*50*9*8

    20000 and 100 configuration:(high because of loop order)
        origin overlap quicktest tight
        5.6s   4.6s     4.6s     18.14s

    merge the first streaming together






## [0.1.4] 2017 April 17
#### if not using -O3
origin: 35.47
overlapped/quicktest: almost the same
move to cluster

#### changed:
    loop order changed to iY->iX



reason: three lines are required each time.
how about proposed method: three lines are requried, but can calculate two iterations

## [0.1.3] 2017 April 12
#### to do
- cllision-streaming-collision-streaming forcely.
#### added
- will be updated in branch: quicktest.
    implemented in collision_streaming_twice().
#### results
-   still 5.6s(is this the bound)?

## [0.1.2] 2017 April 11
#### added
- cmake support
#### changed
- add -O3 also in link flag(default is only in comple flag),
#### problems
- if run all tasks in one scripts, latter case use more time.
- results not stable
#### done
verify results of two experiements

## [0.1.1] 2017 Apil 10
#### added
- now streaming is inside collision
- 5.6s for 10000 steps

## [0.1.0] 2017 Apil 10
#### added
- timer added
- save disabled
- progagate will be inside colision, and be put into tmpbuffer.
- 6.3 s for 10000 steps




lb.h
/*  Lattice Boltzmann sample, written in C
 *
 *  Copyright (C) 2006 Jonas Latt
 *  Address: Rue General Dufour 24,  1211 Geneva 4, Switzerland
 *  E-mail: Jonas.Latt@cui.unige.ch
 *
 *  This program is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU General Public License
 *  as published by the Free Software Foundation; either version 2
 *  of the License, or (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public
 *  License along with this program; if not, write to the Free
 *  Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
 *  Boston, MA  02110-1301, USA.
 */

/* lb.h: Assembly of various functions for simulating a lattice
 *       Boltzmann dynamics with BGK collision term on a two-
 *       dimenional D2Q9 lattice. The code is generic: every
 *       lattice site can have a different dynamics.
 *       Other collision terms than BGK can easily be added.
 */

#ifndef LB_H
#define LB_H

#include "stddef.h"

#define ZGB

#ifdef _OPENMP
#include <omp.h>
#endif

#define NUM_EVENTS 4 
#ifdef ADDPAPI
#include "papi.h"
static int EventSet[] = {PAPI_L1_TCM, PAPI_L2_TCM, PAPI_L3_TCM,
                          PAPI_TLB_DM, //PAPI_TLB_IM,
                          //PAPI_L2_ICA, PAPI_L3_ICA,
                          // PAPI_L2_DCA, PAPI_L3_DCA,
                          //PAPI_L1_DCM, PAPI_L1_ICM,
                          //PAPI_L2_DCM, PAPI_L2_ICM
                          };
#endif

/* struct Dynamics                                               */
/*****************************************************************/
/*   emulation of a class that defines the dynamics of a lattice
 *   site. The function dynamicsFun contains the algorithm of the
 *   collision, and selfData points to the function arguments
 *   (for example the local viscosity)
 */
typedef struct {
    void   (*dynamicsFun) (double* fPop, void* selfData);
    void* selfData;
} Dynamics;


/* struct Node                                                   */
/*****************************************************************/
/*  a lattice node, containing the data (distribution functions)
 *  for a D2Q9 lattice, and a pointer to the local dynamics, i.e.
 *  the collision term. Two "methods" are added to construct and
 *  initialize a node.
 */
typedef struct {
    double    fPop[9];
    Dynamics* dynamics;
} Node;

void constructNode(Node* node);
void iniEquilibrium(Node* node, double rho, double ux, double uy);


/* struct Simulation                                             */
/*****************************************************************/
/*  a full D2Q9 LB simulation, containing the allocated memory
 *  for the lattice. Some "methods" are added to initiate the
 *  dynamics.
 */
typedef struct {
    int lx, ly;               // lx*ly lattice
    Node*  memoryChunk;       // contiguous raw memory
    Node*  tmpMemoryChunk;    // contiguous raw tmp memory
    Node*  tmpMemoryChunk2;
    Node** lattice;           // lattice, points to raw memory
    Node** tmpLattice;        // tmp lasttice, points to raw memory
} Simulation;

/* added by lifeng & Yuankun Fu                                */
/*****************************************************************/
void updateZeroGradientBoundary();

void collide_with_stream(Simulation* sim);
void collide_with_stream_openmp(Simulation* sim);
void collide_with_stream_twice(Simulation* sim);
void finalize_stream(Simulation* sim);

void collide_tight(Simulation* sim);
void collide_tight2(Simulation* sim);
void collide_tight_openmp(Simulation* sim);

void collide_tight_block(Simulation* sim);
void collide_tight_block_openmp(Simulation* sim);

void collide_tight_panel_ix(Simulation* sim);
void collide_tight_panel_iy(Simulation* sim);
void collide_tight_panel_iy_openmp(Simulation* sim);
void finalize_tight(Simulation* sim);

void constructSim(Simulation* sim, int lx, int ly);
void destructSim(Simulation* sim);
void setDynamics(Simulation* sim, int iX, int iY, Dynamics* dyn);

void collide(Simulation* sim);
void collide_openmp(Simulation* sim);
void propagate(Simulation* sim);
void propagate_openmp(Simulation* sim);

void makePeriodic(Simulation* sim);
void saveVel(Simulation* sim, char fName[]);
void saveF(Simulation* sim, int iPop, char fName[]);


/* some free helper functions                                    */
/*****************************************************************/

  // compute density and velocity from the f's
void computeMacros(double* f, double* rho, double* ux, double* uy);

  // compute local equilibrium from rho and u
double computeEquilibrium(int iPop, double rho,
                          double ux, double uy, double uSqr);
  // bgk collision term
void bgk(double* fPop, void* selfData);

#endif


run_exps.sh
#!/bin/bash
WORK_DIR=`pwd`
RESULT_DIR=${WORK_DIR}/results/`date +"%y%m%d_%H%M"`
mkdir -pv ${RESULT_DIR}

BUILD_DIR=${WORK_DIR}/build/bin
cd ${RESULT_DIR}


#dimension
# smallest basic test
# lx=8
# ly=8
# export STOP=10 #time stop
# thread_block=2 #each thread will compute this number of x lines or y lines
# blk_size=2

# lx=16
# ly=16
# export STOP=10 #time stop
# thread_block=4 #each thread will compute this number of x lines or y lines
# export OMP_NUM_THREADS=4
# blk_size=2

# lx=64
# ly=64
# export STOP=10 #time stop
# thread_block=16 #each thread will compute this number of x lines or y lines
# export OMP_NUM_THREADS=4
# blk_size=4

# verifiction test
lx=256
ly=256
export STOP=100 #time stop
thread_block=64 #each thread will compute this number of x lines or y lines
blk_size=32
export OMP_NUM_THREADS=4
export KMP_AFFINITY=verbose, granularity=fine, compact
#blk_size=(1 2 4 8 16 32 64 128 256 512)
##for CASE_NAME in tight_block_papi
#for CASE_NAME in tight_block
#do
    #for ((k=0; k<${#blk_size[@]}; k++))
    #do
       #for i in `seq 3`
        #do
            #${BUILD_DIR}/unsteady_${CASE_NAME} $lx $ly ${blk_size[k]}

            #echo "No.$i exp of case ${CASE_NAME} done "
        #done
    #done
#done


# export STOP=5000 #time stop
# thread_block=16 #each thread will compute this number of x lines or y lines
# blk_size=4
# export OMP_NUM_THREADS=16

# for CASE_NAME in  origin combine quicktest tight tight_block segment
# for CASE_NAME in origin combine tight

# for CASE_NAME in origin tight origin_openmp combine_openmp tight_openmp tight_block_openmp panel panel_openmp
# for CASE_NAME in origin_openmp_papi tight_openmp_papi tight_block_openmp_papi
# for CASE_NAME in origin combine tight origin_openmp combine_openmp tight_openmp origin_papi origin_openmp_papi combine_papi combine_openmp_papi tight_papi tight_openmp_papi tight_block tight_block_openmp tight_block_papi tight_block_openmp_papi panel panel_openmp panel_papi panel_openmp_papi
for CASE_NAME in origin_papi tight_block_openmp_papi
do
    echo "result will be saved in ${RESULT_DIR}"
    for i in `seq 1`
    do
        # ${BUILD_DIR}/unsteady $lx $ly $blk_size &> ${RESULT_DIR}/${CASE_NAME}_${i}.log
        ${BUILD_DIR}/unsteady_${CASE_NAME} $lx $ly $blk_size $thread_block &>> ${RESULT_DIR}/log
        echo "No.$i exp of case ${CASE_NAME} done " &>> ${RESULT_DIR}/log
    done
done

# blk_size=32
# chunk_size=32
# for CASE_NAME in segment segment_openmp
# do
#     echo "result will be saved in ${RESULT_DIR}"
#     for i in `seq 1`
#     do
#         # ${BUILD_DIR}/unsteady $lx $ly $blk_size &> ${RESULT_DIR}/${CASE_NAME}_${i}.log
#         ${BUILD_DIR}/unsteady_${CASE_NAME} $lx $ly $blk_size $chunk_size &>> ${RESULT_DIR}/log
#         echo "No.$i exp of case ${CASE_NAME} done " &>> ${RESULT_DIR}/log
#     done
# done

#run tight_block with different block size
# blk_size=(32)
# for CASE_NAME in tight_block tight_block_openmp
# do
#     echo "result will be saved in ${RESULT_DIR}"
#     for ((k=0; k<${#blk_size[@]}; k++)); do
#         for i in `seq 1`
#         do
#             ${BUILD_DIR}/unsteady_${CASE_NAME} $lx $ly ${blk_size[k]} &>> ${RESULT_DIR}/log
#             echo "No.$i exp of case ${CASE_NAME} ${blk_size[k]} done " &>> ${RESULT_DIR}/log
#         done
#     done
# done

# #run tight_block with different segment size
# blk_size=(32)
# for CASE_NAME in segment segment_openmp
# do
#     echo "result will be saved in ${RESULT_DIR}"
#     for ((k=0; k<${#blk_size[@]}; k++)); do
#         for i in `seq 1`
#         do
#             ${BUILD_DIR}/unsteady_${CASE_NAME} $lx $ly ${blk_size[k]} &>> ${RESULT_DIR}/log
#             echo "No.$i exp of case ${CASE_NAME} ${blk_size[k]} done " &>> ${RESULT_DIR}/log
#         done
#     done
# done


unsteady.c
/*  Lattice Boltzmann sample, written in C
 *
 *  Copyright (C) 2006 Jonas Latt
 *  Address: Rue General Dufour 24,  1211 Geneva 4, Switzerland
 *  E-mail: Jonas.Latt@cui.unige.ch
 *
 *  This program is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU General Public License
 *  as published by the Free Software Foundation; either version 2
 *  of the License, or (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public
 *  License along with this program; if not, write to the Free
 *  Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
 *  Boston, MA  02110-1301, USA.
*/

/* unsteady.c:
 * This example examines an unsteady flow past a cylinder placed in a channel.
 * The cylinder is offset somewhat from the center of the flow to make the
 * steady-state symmetrical flow unstable. At the inlet, a Poiseuille profile is
 * imposed on the velocity, where as the outlet implements an outflow condition:
 * grad_x u = 0. At Reynolds numbers around 100, an unstable periodic pattern is
 * created, the Karman vortex street.
 */

#include "lb.h"
#include "boundaries.h"
#include <stdio.h>
#include <stdlib.h>
#include "eval_tools.h"

/*#define NO_SAVE*/

  // These constants define the flow geometry and are commented in
  //   the function setConstants()
int lx, ly;
int obst_x, obst_y, obst_r;
double uMax, Re, nu, omega;
int maxT, tSave;
int NUM_THREADS;

/********************* added by Yuankun Fu ***********************/
int blk_size;
double *myrho1, *myrho2;
int iT=0, count;
int thread_block;
// #define TIGHT2

/*#ifdef ADDPAPI*/
  // int EventSet[] = {PAPI_TLB_DM, PAPI_L1_DCM, PAPI_L2_DCM, PAPI_L3_DCM};
  long long global_CM[NUM_EVENTS];
  // int EventSet[] = {PAPI_L1_DCM, PAPI_L2_DCM, PAPI_TLB_DM};
/*#endif*/
/********************* ******************** ***********************/

  // The dynamics that are to be specified on different regions of
  //   the flow: LBGK in the bulk, bounce-back on the cylinder,
  //   regularized boundary condition on the four domain boundaries.
  //   Alternatively, the Zou/He boundary condition can be used:
  //   replace upperRegularized by upperZouHe etc.
Dynamics       bulkDynamics         = { &bgk, (void*)&omega };
Dynamics       bounceBackDynamics   = { &bounceBack, 0 };

  // The velocity to be imposed on the upper/lower boundary: u=0
VelocityBCData zeroVelocityBoundary = { &bulkDynamics, 0., 0. };

Dynamics upperBoundary = { &upperRegularized,
                           (void*)&zeroVelocityBoundary };
Dynamics lowerBoundary = { &lowerRegularized,
                           (void*)&zeroVelocityBoundary };
Dynamics* leftBoundary;   // Those two objects are initialized in the
Dynamics* rightBoundary;  //   function iniData()

  // These arrays contain the velocities that are to be imposed on
  //   the inlet (poiseuilleBoundary) and the outlet
  //   (pressureBoundary) of the channel
VelocityBCData* poiseuilleBoundary;
PressureBCData* pressureBoundary;

  // The main object containing the simulation
Simulation sim;

void setConstants(int argc, char *argv[]) {
    /*lx       = 250;     // channel length
    ly       = 50;      // channel height
    obst_r = lx/10+1;   // radius of the cylinder
    obst_x = lx/5;
    obst_y = ly/2;*/
    
    lx = atoi(argv[1]);     //channel lenghth
    ly = atoi(argv[2]);     //channel width
    blk_size = atoi(argv[3]);

    obst_r = 8;   // radius of the cylinder
    obst_x = lx/4;      // position of the cylinder; the cylinder is
    obst_y = ly/2;      // offset from the center to break symmetry

    uMax  = 0.02;       // maximum velocity of the Poiseuille inflow
    Re    = 100;        // Reynolds number
    nu    = uMax * 2.*obst_r / Re;  // kinematic fluid viscosity
    omega = 1. / (3*nu+1./2.);      // relaxation parameter

    maxT   = atoi(getenv("STOP"));       // total number of iterations
    tSave  = 2;          // frequency of periodic saves to disk

    printf("\nlx=%d, ly=%d, omega=%f, blk_size=%d\n\n", lx, ly, omega, blk_size);
}

  // Memory allocation and default initialisation of the simulation
  //   and the left/right boundaries
void iniData() {
    int iX, iY;
    poiseuilleBoundary =
        (VelocityBCData*) calloc(ly+2, sizeof(VelocityBCData));
    pressureBoundary =
        (PressureBCData*) calloc(ly+2, sizeof(PressureBCData));

    leftBoundary  = (Dynamics*) calloc(ly+2, sizeof(Dynamics));
    rightBoundary = (Dynamics*) calloc(ly+2, sizeof(Dynamics));

#ifdef ZGB
    //add by Yuankun
    myrho1 = (double *) calloc(ly, sizeof(double));
    myrho2 = (double *) calloc(ly, sizeof(double));
#endif

    for (iY=2; iY<=ly-1; ++iY) {
        poiseuilleBoundary[iY].bulkDynamics   = &bulkDynamics;
        poiseuilleBoundary[iY].uy             = 0.;
        pressureBoundary[iY].bulkDynamics = &bulkDynamics;
        pressureBoundary[iY].rho          = 1.;
        pressureBoundary[iY].uPar         = 0.;

        leftBoundary[iY].dynamicsFun = &leftRegularized;
        leftBoundary[iY].selfData
            = (void*)&poiseuilleBoundary[iY];
        rightBoundary[iY].dynamicsFun = &rightPressureRegularized;
        rightBoundary[iY].selfData
            = (void*)&pressureBoundary[iY];
    }
}

  // De-allocation of the memory
void freeData() {
    free(rightBoundary);
    free(leftBoundary);
    free(poiseuilleBoundary);
    free(pressureBoundary);

    //add by Yuankun
    free(myrho1);
    free(myrho2);
}

  // compute parabolic Poiseuille profile
double computePoiseuille(int iY) {
    double y = (double)(iY-1);
    double L = (double)(ly-1);
    return 4.*uMax / (L*L) * (L*y-y*y);
}

  // Specify the geometry of the simulation
void iniGeometry() {
    int iX, iY;
    for(iX=1; iX<=lx; ++iX) {
        for(iY=1; iY<=ly; ++iY) {
              // All bulk nodes are initialized at equilibrium with constant
              // density and a velocity determined by a y-dependend Poiseuille
              // profile.
            double uPoiseuille = computePoiseuille(iY);
            iniEquilibrium(&sim.lattice[iX][iY], 1., uPoiseuille, 0.);
              // on the obstacle, set bounce back dynamics
            if ( (iX-obst_x)*(iX-obst_x) +
                 (iY-obst_y)*(iY-obst_y) <= obst_r*obst_r )
            {
                setDynamics(&sim, iX, iY, &bounceBackDynamics);
            }
            //   //add by Yuankun
            // else if ( (iY==1) || (iY==ly) ){
            //     setDynamics(&sim, iX, iY, &bounceBackDynamics);
            // }
              // elsewhere, use lbgk dynamics
            else {
                setDynamics(&sim, iX, iY, &bulkDynamics);
            }
        }
    }

      // upper and lower boundary: u=0
    for (iX=1; iX<=lx; ++iX) {
        setDynamics(&sim, iX, 1, &lowerBoundary);
        setDynamics(&sim, iX, ly, &upperBoundary);
    }

      // left boundary: Poiseuille profile, constant through time
      // right boundary: initially Poiseuille profile, then outlet
      //   condition grad_x u = 0
    for (iY=2; iY<=ly-1; ++iY) {
        double uPoiseuille = computePoiseuille(iY);
        poiseuilleBoundary[iY].ux   = uPoiseuille;
        setDynamics(&sim, 1, iY, &leftBoundary[iY]);
        setDynamics(&sim, lx, iY, &rightBoundary[iY]);
    }
}

  // Compute a second order extrapolation on the right boundary to
  // ensure a zero-gradient boundary condition on the pressure.
  // This must be recomputed at every time step. The velocity is
  // constrained to be perpendicular to the outflow surface.
void updateZeroGradientBoundary() {
    int iY;
    double rho1, ux1, uy1, rho2, ux2, uy2;
    for (iY=2; iY<=ly-1; ++iY) {
        computeMacros(sim.lattice[lx-1][iY].fPop, &rho1, &ux1, &uy1);
        computeMacros(sim.lattice[lx-2][iY].fPop, &rho2, &ux2, &uy2);
        pressureBoundary[iY].rho = 4./3.*rho1 - 1./3.*rho2;
        pressureBoundary[iY].uPar = 0.; //uy=0

        // if(iT==1 || iT==2){
        //     printf("in ZGB, iT=%d, rho2[%d]=%f, rho1[%d]=%f\n", iT, iY, rho2, iY, rho1);
        //     fflush(stdout);
        // }
    }
}

static const double t[9] = { 4./9., 1./9., 1./9., 1./9., 1./9.,
                             1./36., 1./36., 1./36., 1./36. };
  // lattice velocities
static const int c[9][2] = {
    {0,0},
    {1,0}, {0,1}, {-1,0}, {0,-1},
    {1,1}, {-1,1}, {-1,-1}, {1,-1}
};

//OpenMP test hello
void Hello(void){

#ifdef _OPENMP
  int my_rank = omp_get_thread_num();
  int thread_count = omp_get_num_threads();
#else
  int my_rank = 0;
  int thread_count = 1;
#endif

  printf("Hello from thread %d of %d\n", my_rank, thread_count);
}



int main(int argc, char *argv[]) {
    int i;
    char case_name[80], filename[80];
    void (*stream_func)(Simulation *)=NULL;
    void (*collision_func)(Simulation *)=NULL;

      // initialisation of a lx * ly simulation
    setConstants(argc, argv);
    int interval = maxT/10;
    // int interval = 1;


#ifdef _OPENMP
    NUM_THREADS = atoi(getenv("OMP_NUM_THREADS"));
    // chunk_size=(lx<ly?lx:ly)/NUM_THREADS;
    // chunk_size = atoi(argv[4]);
    // thread_block = lx/NUM_THREADS;
    thread_block = atoi(argv[4]);
    printf("thread_block=%d\n", thread_block);
    omp_set_num_threads(NUM_THREADS);
#endif

#ifdef _OPENMP
#pragma omp parallel
#endif
    Hello();

#if defined(COMBINE)
        sprintf(case_name, "combine");
        #ifdef ADDPAPI
        sprintf(case_name, "combine_papi");
        #endif
        collision_func=&collide_with_stream;

  #ifdef _OPENMP
        sprintf(case_name, "combine_openmp");
        #ifdef ADDPAPI
        sprintf(case_name, "combine_openmp_papi");
        #endif
        collision_func=&collide_with_stream_openmp;
  #endif
        stream_func=&finalize_stream;

#elif defined(QUICKTEST)
    #define TWOSTEPS
        sprintf(case_name, "quicktest");
        collision_func=&collide_with_stream_twice;
        stream_func=&finalize_stream;

#elif defined(TIGHT)
    #define TWOSTEPS
        sprintf(case_name, "tight");
        #ifdef ADDPAPI
        sprintf(case_name, "tight_papi");
        #endif
        collision_func=&collide_tight;

    #ifdef _OPENMP
        sprintf(case_name, "tight_openmp");
        #ifdef ADDPAPI
        sprintf(case_name, "tight_openmp_papi");
        #endif
        collision_func=&collide_tight_openmp;
    #endif

    #ifdef TIGHT2
        sprintf(case_name, "tight2");
        collision_func=&collide_tight2;
    #endif
        // no need to swap!
        stream_func=&finalize_stream;

#elif defined(TIGHT_BLOCK)
    #define TWOSTEPS
        sprintf(case_name, "tight_block");
        #ifdef ADDPAPI
        sprintf(case_name, "tight_block_papi");
        #endif
        collision_func=&collide_tight_block;
    #ifdef _OPENMP
        sprintf(case_name, "tight_block_openmp");
        #ifdef ADDPAPI
        sprintf(case_name, "tight_block_openmp_papi");
        #endif
        collision_func=&collide_tight_block_openmp;
    #endif
        // no need to swap!
        stream_func=&finalize_stream;

#elif defined(PANEL)
    #define TWOSTEPS
        sprintf(case_name, "panel");
        #ifdef ADDPAPI
        sprintf(case_name, "panel_papi");
        #endif
        // collision_func=&collide_tight_panel_ix;
        collision_func=&collide_tight_panel_iy;
    #ifdef _OPENMP
        sprintf(case_name, "panel_openmp");
        #ifdef ADDPAPI
        sprintf(case_name, "panel_openmp_papi");
        #endif
        collision_func=&collide_tight_panel_iy_openmp;
    #endif
        // no need to swap!
        stream_func=&finalize_stream;

#else
        sprintf(case_name, "origin");
        #ifdef ADDPAPI
        sprintf(case_name, "origin_papi");
        #endif
        collision_func=&collide;
        stream_func=&propagate;
    #ifdef _OPENMP
        sprintf(case_name, "origin_openmp");
        #ifdef ADDPAPI
        sprintf(case_name, "origin_openmp_papi");
        #endif
        collision_func=&collide_openmp;
        stream_func=&propagate_openmp;
    #endif

#endif

    printf("use case %s\n", case_name);
    fflush(stdout);

    // allocate spaces for boundries
    iniData();
    // printf("Pass iniData\n");
    // fflush(stdout);

    // allocae space for nodes and lattice
    // set configurations
    constructSim(&sim, lx, ly);
    // printf("Pass constructSim\n");
    // fflush(stdout);

    // bounce back dynamics in obstacles, else use lbgk(bulk dynamics)
    iniGeometry();
    printf("Pass iniGeometry, obst_x=%d, obst_y=%d, obst_r=%d\n", 
             obst_x, obst_y, obst_r);
    fflush(stdout);

    double t_start, t_end, t0, t1, t_collid=0.0, t_zgb=0.0, t_stream=0.0;
    t_start = get_cur_time();

#ifdef ADDPAPI
  int retval = PAPI_library_init (PAPI_VER_CURRENT);
  if (retval != PAPI_VER_CURRENT) {
    printf("PAPI_library_init error!\n");
    exit(1);
  }

  int thread_count;
  #ifdef _OPENMP
    if (PAPI_thread_init((unsigned long (*)(void))(omp_get_thread_num)) != PAPI_OK){
      printf("PAPI_thread_init error!\n");
      exit(1);
    }

    thread_count = NUM_THREADS;
  #else
    thread_count = 1;
  #endif
  // long long realtime[2];

  // realtime[0] = 0;
  // realtime[1] = 0;

  for(i=0; i<NUM_EVENTS; i++)
      global_CM[i] = 0;
  // retval = PAPI_create_eventset(EventSet);
  // if (PAPI_create_eventset(&EventSet) != PAPI_OK){
  //   printf("PAPI library create error!\n");
  //   exit(1);
  // }

#endif

      // the main loop over time steps
    // int iT, count;
    count =0;
#ifndef TWOSTEPS
    for (iT=0; iT<maxT; iT+=1) {
#else
    for (iT=0; iT<maxT; iT+=2) {
#endif
        if (iT%interval==0) {
            printf("t=%d\n", iT);
            fflush(stdout);
        }

        //save before computing
/*#ifndef NO_SAVE*/
        /*if (iT%tSave==0) {*/
            /*printf("iT=%d, save, count=%d\n", iT, count);*/
            /*fflush(stdout);*/
            /*sprintf(filename, "vel_%s_%d.dat", case_name, count);*/
            /*saveVel(&sim, filename);*/
            /*count++;*/
        /*}*/
/*#endif*/

#ifdef ZGB
        t0 = get_cur_time();
        // on the right boundary, outlet condition grad_x u = 0
        updateZeroGradientBoundary();
        t1 = get_cur_time();
        t_zgb += t1-t0;
#endif

        //save after ZGB
/*#ifndef NO_SAVE*/
        /*if (iT%tSave==0) {*/
            /*printf("iT=%d, save, count=%d\n", iT, count);*/
            /*fflush(stdout);*/
            /*sprintf(filename, "vel_%s_zgb_%d.dat", case_name, iT);*/
            /*saveVel(&sim, filename);*/
        /*}*/
/*#endif*/

        // step 3,4,5:compute rho, u, get f^eq, then update f.
        t0 = get_cur_time();
        collision_func(&sim);
        t1 = get_cur_time();
        t_collid += t1-t0;


//           // the data are written to disk after collision, to be
//           //   that the macroscopic variables are computed
//           //   correctly on the boundaries

/* #ifndef NO_SAVE*/
         /*if (iT%tSave==0) {*/
             /*sprintf(filename, "vel_%s_collide_%d.dat", case_name, iT);*/
             /*saveVel(&sim, filename);*/
         /*}*/

 /*#endif*/

        // step 2: streamming step
#if defined(TIGHT)
        //do nothing
#elif defined(TIGHT_BLOCK)
        //do nothing
#elif defined(PANEL)
        //do nothing
#else
        t0 = get_cur_time();
        stream_func(&sim);
        t1 = get_cur_time();
        t_stream += t1-t0;
#endif

        //save after stream
#ifndef NO_SAVE
        if (iT%tSave==0) {
            printf("iT=%d, save, count=%d\n", iT, count);
            fflush(stdout);
            sprintf(filename, "vel_%s_stream_%d.dat", case_name, count);
            /*sprintf(filename, "vel_%s_stream_%d.dat", case_name, iT);*/
            saveVel(&sim, filename);
            count++;
        }
#endif


          // By default: periodic boundary conditions. In this case,
          //   this is important, because the upper and lower
          //   boundaries are horizontally periodic, so that no
          //   special corner nodes are needed.
// #if defined(TIGHT)
//         //do nothing
// #else
        // makePeriodic(&sim);
// #endif

          // the data are written to disk after streaming and Periodic, to be
          //   that the macroscopic variables are computed
          //   correctly on the boundaries


/*
        updateZeroGradientBoundary();
        collision_func(&sim);
        stream_func(&sim);
        makePeriodic(&sim);
        */
    }

#ifdef ADDPAPI
  /* Collect the data into the variables passed in */
    for(i=0; i<NUM_EVENTS; i++){
      //Event[%d]
      printf("Global_E[%02d]: %lld\n", i, global_CM[i]);
    }
    fflush(stdout);

    printf("AE: ");
    for(i=0; i<NUM_EVENTS; i++){
    //Avg_Event[%d]
    printf("%lld ", global_CM[i]/thread_count);
    }
    printf("\n");
    fflush(stdout);

#endif

    //save final data
#ifndef NO_SAVE
    sprintf(filename, "vel_%s_%d.dat", case_name, count);
    saveVel(&sim, filename);
#endif

    t_end = get_cur_time();
    printf("time eclapsed in %d steps, t_total=%.3f, t_collid=%.3f, t_stream=%.3f, t_zgb=%.3f\n",
      maxT, t_end-t_start, t_collid, t_stream, t_zgb);



    destructSim(&sim);
    freeData();
}
